### ISSUE TITLE
Seldom exceptions with Alpine and multiple container layers.

### ISSUE THREAD

**Issue by sdurrheimer** [pires/kubernetes-elasticsearch-cluster#89]
Hi there,

I'm unable to run any elasticsearch pod of this setup, I keep facing the following error:

```
[2017-04-10T14:45:05,518][INFO ][o.e.n.Node               ] [es-master-3871649794-ndrl3] initializing ...
[2017-04-10T14:45:05,565][WARN ][o.e.b.ElasticsearchUncaughtExceptionHandler] [es-master-3871649794-ndrl3] uncaught exception in thread [main]
org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: failed to obtain node locks, tried [[/data/data/elasticsearch-logging]] with lock id [0]; maybe these locations are not writable or multiple nodes were started without increasing [node.max_local_storage_nodes] (was [1])?
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:127) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:114) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:58) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:122) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.cli.Command.main(Command.java:88) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:91) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:84) ~[elasticsearch-5.3.0.jar:5.3.0]
Caused by: java.lang.IllegalStateException: failed to obtain node locks, tried [[/data/data/elasticsearch-logging]] with lock id [0]; maybe these locations are not writable or multiple nodes were started without increasing [node.max_local_storage_nodes] (was [1])?
	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:260) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.node.Node.<init>(Node.java:258) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.node.Node.<init>(Node.java:238) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap$6.<init>(Bootstrap.java:242) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:242) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:360) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:123) ~[elasticsearch-5.3.0.jar:5.3.0]
	... 6 more
Caused by: java.io.IOException: failed to obtain lock on /data/data/nodes/0
	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:239) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.node.Node.<init>(Node.java:258) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.node.Node.<init>(Node.java:238) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap$6.<init>(Bootstrap.java:242) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:242) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:360) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:123) ~[elasticsearch-5.3.0.jar:5.3.0]
	... 6 more
Caused by: java.io.IOException: Mount point not found
	at sun.nio.fs.LinuxFileStore.findMountEntry(LinuxFileStore.java:91) ~[?:?]
	at sun.nio.fs.UnixFileStore.<init>(UnixFileStore.java:65) ~[?:?]
	at sun.nio.fs.LinuxFileStore.<init>(LinuxFileStore.java:44) ~[?:?]
	at sun.nio.fs.LinuxFileSystemProvider.getFileStore(LinuxFileSystemProvider.java:51) ~[?:?]
	at sun.nio.fs.LinuxFileSystemProvider.getFileStore(LinuxFileSystemProvider.java:39) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.getFileStore(UnixFileSystemProvider.java:368) ~[?:?]
	at java.nio.file.Files.getFileStore(Files.java:1461) ~[?:1.8.0_121]
	at org.elasticsearch.env.ESFileStore.getMatchingFileStore(ESFileStore.java:107) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.env.Environment.getFileStore(Environment.java:340) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.env.NodeEnvironment$NodePath.<init>(NodeEnvironment.java:108) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:227) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.node.Node.<init>(Node.java:258) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.node.Node.<init>(Node.java:238) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap$6.<init>(Bootstrap.java:242) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:242) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:360) ~[elasticsearch-5.3.0.jar:5.3.0]
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:123) ~[elasticsearch-5.3.0.jar:5.3.0]
	... 6 more
```

I also tried with a simple `docker run` on my local machine, same problem:

```
docker run --rm -t -i --name elasticsearch \
        --privileged \
        --volume $(pwd)/data:/data \
        -e CLUSTER_NAME=myclustername \
        quay.io/pires/docker-elasticsearch:5.3.0
```

**Comment by pires** [comment]
`java.io.IOException: Mount point not found` means `$(pwd)/data` doesn't exist.

**Comment by sdurrheimer** [comment]
Docker auto-create a folder mounted as a volume so the data directory exists, but even without the volume the docker image start with the same error. There is clearly a problem.

**Comment by pires** [comment]
Are you running this on a Mac? If yes, what's your setup?

**Comment by sdurrheimer** [comment]
Linux 16.04 with Docker 17.04.0-ce. Also tested with Docker 1.12.6.

**Comment by pires** [comment]
Can you create `data` folder and run from there? Seems to work here.

```
pires ~ $ dm ssh dev
                        ##         .
                  ## ## ##        ==
               ## ## ## ## ##    ===
           /"""""""""""""""""\___/ ===
      ~~~ {~~ ~~~~ ~~~ ~~~~ ~~~ ~ /  ===- ~~~
           \______ o           __/
             \    \         __/
              \____\_______/
 _                 _   ____     _            _
| |__   ___   ___ | |_|___ \ __| | ___   ___| | _____ _ __
| '_ \ / _ \ / _ \| __| __) / _` |/ _ \ / __| |/ / _ \ '__|
| |_) | (_) | (_) | |_ / __/ (_| | (_) | (__|   <  __/ |
|_.__/ \___/ \___/ \__|_____\__,_|\___/ \___|_|\_\___|_|
Boot2Docker version 17.03.0-ce, build HEAD : f11a204 - Thu Mar  2 00:14:47 UTC 2017
Docker version 17.03.0-ce, build 3a232c8
docker@dev:~$ ls
docker@dev:~$ pwd
/home/docker
docker@dev:~$ mkdir data
docker@dev:~$ docker run --rm -t -i --name elasticsearch         --privileged         --volume $(pwd)
/data:/data         -e CLUSTER_NAME=myclustername         quay.io/pires/docker-elasticsearch:5.3.0
[2017-04-11T12:03:01,547][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] initializing ...
[2017-04-11T12:03:01,677][INFO ][o.e.e.NodeEnvironment    ] [7258d92a-07c1-40d3-958f-e5479db66411] using [1] data paths, mounts [[/data (tmpfs)]], net usable_space [3.2gb], net total_space [3.4gb], spins? [no], types [tmpfs]
[2017-04-11T12:03:01,679][INFO ][o.e.e.NodeEnvironment    ] [7258d92a-07c1-40d3-958f-e5479db66411] heap size [495.3mb], compressed ordinary object pointers [true]
[2017-04-11T12:03:01,680][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] node name [7258d92a-07c1-40d3-958f-e5479db66411], node ID [KX7Ur49zT3aPFDSTxnbfxg]
[2017-04-11T12:03:01,681][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] version[5.3.0], pid[15], build[3adb13b/2017-03-23T03:31:50.652Z], OS[Linux/4.4.52-boot2docker/amd64], JVM[Oracle Corporation/OpenJDK 64-Bit Server VM/1.8.0_121/25.121-b13]
[2017-04-11T12:03:02,758][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [aggs-matrix-stats]
[2017-04-11T12:03:02,758][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [ingest-common]
[2017-04-11T12:03:02,764][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [lang-expression]
[2017-04-11T12:03:02,764][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [lang-groovy]
[2017-04-11T12:03:02,764][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [lang-mustache]
[2017-04-11T12:03:02,765][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [lang-painless]
[2017-04-11T12:03:02,765][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [percolator]
[2017-04-11T12:03:02,766][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [reindex]
[2017-04-11T12:03:02,766][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [transport-netty3]
[2017-04-11T12:03:02,766][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] loaded module [transport-netty4]
[2017-04-11T12:03:02,767][INFO ][o.e.p.PluginsService     ] [7258d92a-07c1-40d3-958f-e5479db66411] no plugins loaded
[2017-04-11T12:03:05,456][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] initialized
[2017-04-11T12:03:05,456][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] starting ...
[2017-04-11T12:03:05,578][WARN ][i.n.u.i.MacAddressUtil   ] Failed to find a usable hardware address from the network interfaces; using random bytes: 44:74:8a:4f:9e:2f:6b:16
[2017-04-11T12:03:05,650][INFO ][o.e.t.TransportService   ] [7258d92a-07c1-40d3-958f-e5479db66411] publish_address {172.17.0.2:9300}, bound_addresses {172.17.0.2:9300}
[2017-04-11T12:03:05,658][INFO ][o.e.b.BootstrapChecks    ] [7258d92a-07c1-40d3-958f-e5479db66411] bound or publishing to a non-loopback or non-link-local address, enforcing bootstrap checks
ERROR: bootstrap checks failed
max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
[2017-04-11T12:03:05,674][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] stopping ...
[2017-04-11T12:03:05,736][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] stopped
[2017-04-11T12:03:05,739][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] closing ...
[2017-04-11T12:03:05,761][INFO ][o.e.n.Node               ] [7258d92a-07c1-40d3-958f-e5479db66411] closed
```

**Comment by sdurrheimer** [comment]
I tried everything, with or without a volume, with the directory created or not, also tried by setting elasticsearch:elasticsearch owernship on it.

I don't think the volume is the problem. Without a volume, elasticsearch crash with the same error.

**Comment by sdurrheimer** [comment]
I might have a idea, could this be related to `overlay2` docker storage driver ?

**Comment by webeindustry** [comment]
I have this same issue. Nothing seems to be working out right now.

**Comment by pires** [comment]
I just proved above that it works and this repo is *always* tested against a Kubernetes cluster, so it may be related to overlay2, yes.

**Comment by webeindustry** [comment]
I solved this by compressing your docker images into a single one. Issue is 4+ base images with overlay1 or 2.

**Comment by pires** [comment]
As you can see above, it works as is too, so the problem may be something else.

**Comment by sdurrheimer** [comment]
As explained in this [issue](elastic/elasticsearch-docker#44), Elastic did face the same problem with their image, it's a alpine/musl bug.
It seems Elastic solves their problem because the upstream 5.3 docker image is working fine for me.

**Comment by webeindustry** [comment]
@sdurrheimer Your link is broken, but this appears to be the exact issue [https://bugs.alpinelinux.org/issues/7093](url)

**Comment by AtzeDeVries** [comment]
Same issue here. I've recreated the docker-elasticsearch container with base 'openjdk:8-jre' and it runs fine.

**Comment by pires** [comment]
I wonder why this works on CoreOS VMs, arch amd64 (my scenario).

**Comment by pires** [comment]
This was fixed on Alpine 3.6.1 and the base image has been 3.6.2 for some time now, so I consider this issue fixed.
